File: python/树/1110. 删点成林.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def delNodes(self, root: TreeNode, to_delete: List[int]) -> List[TreeNode]:
        s = set(to_delete)
        res = []
        def dfs(root):
            if not root: return 
            cur = root.val
            if cur in s:
                if root.left:
                    if not root.left.val in s:
                        res.append(root.left)
                    dfs(root.left)
                if root.right:
                    if not root.right.val in s:
                        res.append(root.right)
                    dfs(root.right)
                return 
            root.left = dfs(root.left)
            root.right = dfs(root.right)
            return root
        dfs(root)
        if root and not root.val in s:
            res.append(root)
        return res

File: go/tree/1110. 删点成林.go
package tree

import "fmt"

func delNodes(root *TreeNode, to_delete []int) []*TreeNode {
	var dfs func(*TreeNode) *TreeNode
	s := make(map[int]struct{})
	var res []*TreeNode
	for _, v := range to_delete {
		s[v] = struct{}{}
	}
	dfs = func(root *TreeNode) *TreeNode {
		if root == nil {
			return nil
		}
		v := root.Val
		if containValue(s, v) {
			if root.Left != nil {
				if !containValue(s, root.Left.Val) {
					fmt.Println(1)
					res = append(res, root.Left)
				}
				dfs(root.Left)
			}
			if root.Right != nil {
				if !containValue(s, root.Right.Val) {
					fmt.Println(1)
					res = append(res, root.Right)
				}
				dfs(root.Right)
			}
			return nil
		}
		root.Left = dfs(root.Left)
		root.Right = dfs(root.Right)
		return root
	}
	dfs(root)
	if root != nil && !containValue(s, root.Val) {
		res = append(res, root)
	}
	return res
}

func containValue(s map[int]struct{}, v int) bool {
	if _, ok := s[v]; ok {
		return true
	}
	return false
}
